### PR TITLE
[AC-1045] vault timeout action policy

### PR DESCRIPTION
**Pull request by jlf0dev**
## Type of change
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective

This change locks the "Vault timeout action" in settings if the vault timeout policy is set with a specified action. It also changes the callout in settings to include the vault timeout action.

## Code changes

* **file.ext:** Description of what was changed and why

## Screenshots

image

image

## Before you submit
- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

**Review by fedemkr — DISMISSED** [review]

File: src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs [hunk 1]
```diff
@@ -374,7 +381,13 @@ public async Task ApproveLoginRequestsAsync()
 
         public async Task VaultTimeoutActionAsync()
         {
-            var options = _vaultTimeoutActions.Select(o =>
+            if (_vaultTimeoutPolicy != null &&
+                !string.IsNullOrEmpty(_policyService.GetPolicyString(_vaultTimeoutPolicy, PolicyService.TIMEOUT_POLICY_ACTION)))
+            {
+                // do nothing if we have a policy set
+                return;
+            }
```
Comment (fedemkr) [inline src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs:389]: ♻️ Not to do anything here but as part of #2443 refactor, we could add a method to `Policy` called `Has(string key)` so then we could simplify the if to:
`if (_vaultTimeoutPolicy != null && _vaultTimeoutPolicy.Has(PolicyService.TIMEOUT_POLICY_ACTION))`
or
`if (_vaultTimeoutPolicy?.Has(PolicyService.TIMEOUT_POLICY_ACTION) == false)`

**Review by fedemkr — DISMISSED** [review]

File: src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs [hunk 2]
```diff
@@ -117,20 +117,27 @@ public async Task InitAsync()
                     _localizeService.GetLocaleShortTime(lastSync.Value));
             }
 
+            _vaultTimeoutPolicy = null;
+            _vaultTimeoutOptions = VaultTimeoutOptions;
+            _vaultTimeoutActionOptions = VaultTimeoutActionOptions;
+
+            _vaultTimeout = await _vaultTimeoutService.GetVaultTimeout();
+            _vaultTimeoutDisplayValue = _vaultTimeoutOptions.FirstOrDefault(o => o.Value == _vaultTimeout).Key;
```
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs:125]: You've got a good point, the `null` would put the application in a state we don't want. I'll change this to just `First` so we error if we encounter something unexpected, which should never happen unless we mess up.

File: src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs (hunk 1 above)
Comment (jlf0dev) [reply src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs:389]: I think thats a great idea! 👍

File: src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fedemkr) [reply src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs]: 🤔 The only scenario I can think of that could potentially break this is having an old mobile app version and a new policy with a timeout value that doesn't exist in the old app, e.g. adding a policy timeout with value "4", which would be very unlikely AFAIK.
Maybe we could check with product team to see if we should provide a safe check here by finding the nearest timeout/option for the non-found policy value (case on FirstOfDefault -> Default). So in the aforementioned example of "4", we should treat it as "5" which does exist in the current list.
What do you think?

File: src/Core/Services/VaultTimeoutService.cs [hunk 3]
```diff
@@ -271,5 +271,31 @@ public async Task ClearAsync(string userId = null)
...
+
+                // We really shouldn't need to set the value here, but multiple services relies on this value being correct.
+                if (vaultTimeoutAction != action)
+                {
+                    await _stateService.SetVaultTimeoutActionAsync(action, userId);
+                }
+                return action;
+            }
+
+            return vaultTimeoutAction;
```
[the remark this answers is not in the document]
Comment (fedemkr) [reply src/Core/Services/VaultTimeoutService.cs:298]: 1. 🎨 IMO having a `GetAsync(PolicyType type)` is more flexible than having properties for each needed policy type. And also it should be an async call cause we can't use properties; given that if the `_policyCache` is not build up when calling, it needs to be filled and that would take an async call.

2. ♻️  If there is no rush to have this feature available then I'd try to refactor both methods given that QA will test the policy and the vault timeout feature. IMO we're wasting time applying the policy every time we get the value from the `VaultTimeoutService` whereas we could just update the "Vault timeout"/"Vault timeout action" only when the policies are synced (in `PolicyService.Replace()` as you mention) or when the user change their values (and we check the policy in this case as well). So the policy logic should be moved out of the `GetVaultTimeout(...)` / `GetVaultTimeoutAction(...)` methods to the policy sync and to `SetVaultTimeoutOptionsAsync(...)`. 
Thoughts? If this is time sensitive or you feel is out of scope we can create a tech-debt task no problem.

[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/Core/Services/VaultTimeoutService.cs]: 1. Something like this?
```c#
public async Task<Policy> GetFirstAsync(PolicyType type, string userId = null)
{
    var policies = await GetAll(type, userId);
    return policies?.FirstOrDefault();
}
```

2. I went ahead and refactored our policy service to update the stored vault timeout when it's updated from a sync in [f5f8c4c](https://github.com/bitwarden/mobile/pull/2415/commits/f5f8c4c2f6ce1ddb493b9267adbfb774e653a539)

File: src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jlf0dev) [reply src/App/Pages/Settings/SettingsPage/SettingsPageViewModel.cs:124]: Update: Ended up changing it in [758365a](https://github.com/bitwarden/mobile/pull/2415/commits/758365af990ef7caa651f19eca3307aeb66dbeec), take a look!

File: src/Core/Abstractions/IStateService.cs [hunk 4]
```diff
@@ -128,8 +128,8 @@ public interface IStateService
         Task SetEventCollectionAsync(List<EventData> value);
         Task<Dictionary<string, FolderData>> GetEncryptedFoldersAsync(string userId = null);
         Task SetEncryptedFoldersAsync(Dictionary<string, FolderData> value, string userId = null);
-        Task<Dictionary<string, PolicyData>> GetEncryptedPoliciesAsync(string userId = null);
-        Task SetEncryptedPoliciesAsync(Dictionary<string, PolicyData> value, string userId = null);
+        Task<Dictionary<string, PolicyData>> GetPoliciesAsync(string userId = null);
```
Comment (mp-bw) [inline src/Core/Abstractions/IStateService.cs:131]: Are these truly no longer encrypted?  If so is there existing data that requires migration?

**Comment by mp-bw** [comment]
Looks good.  Because it's still fresh in my head, can you confirm `null` is still handled properly re: vault timeout so [this](https://github.com/bitwarden/mobile/pull/2440) doesn't happen again? :)

**Comment by jlf0dev** [comment]
Yep, looks like `null` is working correctly!

**Review by mp-bw — APPROVED** [review]

**Pull request merged**